diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6cabd8e..b678c4e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -45,15 +45,17 @@ foreach (PKG MPC MPFR PIRANHA FLINT LLVM)
     endif()
 endforeach()
 
-message("CMAKE_BUILD_TYPE        : ${CMAKE_BUILD_TYPE}")
-message("CMAKE_CXX_FLAGS         : ${CMAKE_CXX_FLAGS}")
-message("CMAKE_CXX_FLAGS_RELEASE : ${CMAKE_CXX_FLAGS_RELEASE}")
-message("CMAKE_CXX_FLAGS_DEBUG   : ${CMAKE_CXX_FLAGS_DEBUG}")
-message("HAVE_SYMENGINE_MPFR     : ${HAVE_SYMENGINE_MPFR}")
-message("HAVE_SYMENGINE_MPC      : ${HAVE_SYMENGINE_MPC}")
-message("HAVE_SYMENGINE_PIRANHA  : ${HAVE_SYMENGINE_PIRANHA}")
-message("HAVE_SYMENGINE_FLINT    : ${HAVE_SYMENGINE_FLINT}")
-message("HAVE_SYMENGINE_LLVM     : ${HAVE_SYMENGINE_LLVM}")
+message("CMAKE_SYSTEM_PROCESSOR              : ${CMAKE_SYSTEM_PROCESSOR}")
+message("CMAKE_BUILD_TYPE                    : ${CMAKE_BUILD_TYPE}")
+message("CMAKE_CXX_FLAGS                     : ${CMAKE_CXX_FLAGS}")
+message("CMAKE_CXX_FLAGS_RELEASE             : ${CMAKE_CXX_FLAGS_RELEASE}")
+message("CMAKE_CXX_FLAGS_DEBUG               : ${CMAKE_CXX_FLAGS_DEBUG}")
+message("HAVE_SYMENGINE_MPFR                 : ${HAVE_SYMENGINE_MPFR}")
+message("HAVE_SYMENGINE_MPC                  : ${HAVE_SYMENGINE_MPC}")
+message("HAVE_SYMENGINE_PIRANHA              : ${HAVE_SYMENGINE_PIRANHA}")
+message("HAVE_SYMENGINE_FLINT                : ${HAVE_SYMENGINE_FLINT}")
+message("HAVE_SYMENGINE_LLVM                 : ${HAVE_SYMENGINE_LLVM}")
+message("HAVE_SYMENGINE_LLVM_LONG_DOUBLE     : ${HAVE_SYMENGINE_LLVM_LONG_DOUBLE}")
 
 message("Copying source of python wrappers into: ${CMAKE_CURRENT_BINARY_DIR}")
 file(COPY symengine/ DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/symengine)
