From 845168d3683f28781b32577abf4f38ed93f0bc79 Mon Sep 17 00:00:00 2001
From: Isuru Fernando <isuruf@gmail.com>
Date: Mon, 30 Sep 2024 11:38:54 -0500
Subject: [PATCH 2/3] Another cython 3.1 fix

---
 cmake/cython_test.pyx | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/cmake/cython_test.pyx b/cmake/cython_test.pyx
index cb803c6..47f9847 100644
--- a/cmake/cython_test.pyx
+++ b/cmake/cython_test.pyx
@@ -75,8 +75,8 @@ cdef extern from "<symengine/symbol.h>" namespace "SymEngine":
         string get_name() nogil
 
 cdef extern from "<symengine/add.h>" namespace "SymEngine":
-    cdef RCP[Basic] add(RCP[Basic] &a, RCP[Basic] &b) nogil except+
-    cdef RCP[Basic] sub(RCP[Basic] &a, RCP[Basic] &b) nogil except+
+    cdef RCP[Basic] add(RCP[Basic] &a, RCP[Basic] &b) except+ nogil
+    cdef RCP[Basic] sub(RCP[Basic] &a, RCP[Basic] &b) except+ nogil
 
     cdef cppclass Add(Basic):
         void as_two_terms(const Ptr[RCP[Basic]] &a, const Ptr[RCP[Basic]] &b)
-- 
2.44.0

